<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-2.8.1"><span class="toc-item-num">2.8.1&nbsp;&nbsp;</span>Вопрос: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-2.8.2"><span class="toc-item-num">2.8.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-2.8.3"><span class="toc-item-num">2.8.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.8.4"><span class="toc-item-num">2.8.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков



**Цель исследования**

Понять как количество детей и семейное положение влияет на факт погашения кредита в срок.

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов мы получим из файла `/datasets/data.csv`.

 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Общий вывод.

## Обзор данных

Импортируем библиотеку `pandas`

In [2]:
# Импортируем библиотеку pandas
import pandas as pd 

Прочитаем файл `data.csv` из папки `/datasets` и сохраним его в переменной `df`:

In [3]:
# Чтение фала с данными и сохранение в df
df = pd.read_csv('/datasets/data.csv') 

Выведем на экран первый 10 строк таблицы `df`

In [4]:
# Получение первых 10 строк таблицы df
display(df.head(10)) 

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим общую информацию о таблице:

In [5]:
# Получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов:
* с типом данных `float64` — 2 шт.
* с типом данных `int64` — 5 шт.
* с типом данных `object` — 5 шт.

Документация к данным:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

**Выводы**

По названиям столбцов видно что они написаны в соответствии со змеиным регистром. Тут все хорошо.
Но в столбцах `days_employed` и `total_income` есть пропуски. Это нужно будет исправить.
Из 10 полученных строчек видно, что в столбце `days_emloyed` есть отрицательные значения. Так не должно быть. В столбце `education` строки записаны в разном регистре. Это тоже нужно будет поправить.

## Предобработка данных 

### Проверка данных на аномалии и исправления.

In [6]:
# Используем abs функцию, которая измененит отрицательных значений на положительные в столбце days_emloyed
df['days_employed'] = df['days_employed'].abs()

In [7]:
# Получение первых 10 строк таблицы df
display(df.head(10)) 

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [8]:
# Посмотрим уникальные значения в столбце children
display(df['children'].unique()) 

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть отрицательные значения и подозрительное значение в 20 детей, хотя мировой рекорд 87.

In [9]:
# Отфильтруем данные. Посмотрим сколько людей с таким колличеством детей
display(df[df['children'] == 20]) 

children  days_employed  dob_years education  education_id  \
606          20     880.221113         21   среднее             1   
720          20     855.595512         44   среднее             1   
1074         20    3310.411598         56   среднее             1   
2510         20    2714.161249         59    высшее             0   
2941         20    2161.591519          0   среднее             1   
...         ...            ...        ...       ...           ...   
21008        20    1240.257910         40   среднее             1   
21325        20     601.174883         37   среднее             1   
21390        20            NaN         53   среднее             1   
21404        20     494.788448         52   среднее             1   
21491        20     173.954460         27   среднее             1   

         family_status  family_status_id gender income_type  debt  \
606    женат / замужем                 0      M   компаньон     0   
720    женат / замужем                 0      F   компаньон     0   
1074   женат / замужем                 0      F   сотрудник     1   
2510    вдовец / вдова                 2      F   сотрудник     0   
2941   женат / замужем                 0      F   сотрудник     0   
...                ...               ...    ...         ...   ...   
21008  женат / замужем                 0      F   сотрудник     1   
21325  женат / замужем                 0      F   компаньон     0   
21390  женат / замужем                 0      M   компаньон     0   
21404  женат / замужем                 0      M   компаньон     0   
21491  женат / замужем                 0      F   сотрудник     0   

        total_income                                purpose  
606    145334.865002                          покупка жилья  
720    112998.738649                   покупка недвижимости  
1074   229518.537004                  получение образования  
2510   264474.835577  операции с коммерческой недвижимостью  
2941   199739.941398                  на покупку автомобиля  
...              ...                                    ...  
21008  133524.010303                        свой автомобиль  
21325  102986.065978                 профильное образование  
21390            NaN             покупка жилой недвижимости  
21404  156629.683642        операции со своей недвижимостью  
21491  155520.899688     на покупку подержанного автомобиля  

[76 rows x 12 columns]

Видно что таких записей 76. Т.е. 76 людей имеют по 20 детей. Это ошибка,
изменим число 20 на 2.

In [10]:
# Используем abs функцию, которая измененит отрицательных значений на положительные в столбце children
df['children'] = df['children'].abs() 

In [11]:
# Используем функцию loc, которая заменит в колонке children значения 20 на 2
df.loc[(df.children == 20), 'children'] = 2 

In [12]:
# Посмотрим еще раз уникальные значения в столбце children
display(df['children'].unique()) 

array([1, 0, 3, 2, 4, 5])

**Выводы**

Отрицательные и подозрительные значения в столбцах `days_employed` и `children` убраны. Теперь данные смотрятся лучше.

### Заполнение пропусков

Посмотрим какое количество пропусков и в каких колонках:

In [13]:
# Вызываем метод isna, чтобы увидеть, где есть пропуски и посчитаем их методом sum 
display(df.isna().sum()) 

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В столбцах `days_employed` и `total_income` 2174 пустых значения

Посмотрим что это за пустые значения в столбцах `days_employed` и `total_income`:

In [14]:
# Посмотрим 10 пустых значений столбца days_employed используя метод isna и head
display(df[df['days_employed'].isna()].head(10)) 

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [15]:
# Посмотрим 10 пустых значений столбца total_income
display(df[df['total_income'].isna()].head(10)) 

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Видим пустое значение `NaN` он принадлежит к типу `float` с ним можно совершать математические операции

Посмотрим, какую долю составляют пропущенные значения в каждом из столбцов с пропусками

In [16]:
# Создаем переменную df_ratio_days_employed_isna, 
# в которой посчитаем долю пропущенных значений в столбце days_employed
df_ratio_days_employed_isna = df['days_employed'].isna().sum()/len(df['days_employed'])*100 

In [17]:
# Создаем переменную df_ratio_days_total_income, 
# в которой посчитаем долю пропущенных значений в столбце total_income
df_ratio_days_total_income = df['total_income'].isna().sum()/len(df['total_income'])*100 

In [18]:
# Посмотрим долю пропущенных значений в столбце days_employed 
display(f'Доля пропущенных значений в столбце days_employed: {int(df_ratio_days_employed_isna)}%') 

'Доля пропущенных значений в столбце days_employed: 10%'

In [19]:
# Посмотрим долю пропущенных значений в столбце total_income 
display(f'Доля пропущенных значений в столбце total_income: {int(df_ratio_days_total_income)}%') 

'Доля пропущенных значений в столбце total_income: 10%'

В столбцах `days_employed` и `total_income` доля пропущенных значений составила: 10%. Возможно просто не указали свой трудовой стаж работы.

Заполним пропущенные значения медианой. Т.к в выборке есть выборке есть выдающиеся значения. 

In [21]:
# Создадим новую переменную df_days_employed_median и получим медианное значение (с помощью метода .median()),
# в стобце days_employed 
df_days_employed_median = df['days_employed'].median() 

In [22]:
# Создадим новую переменную df_total_income_median и получим медианное значение (с помощью метода .median()),
# в стобце total_income
df_total_income_median = df['total_income'].median() 

In [23]:
# Используем метод fillna() для заполнения пропусков медианой
df['days_employed'].fillna(df_days_employed_median, inplace = True) 

In [24]:
# Используем метод fillna() для заполнения пропусков медианой
df['total_income'].fillna(df_total_income_median, inplace = True) 

In [25]:
# Вызываем метод isna, чтобы увидеть, остались ли пропуски
display(df.isna().sum()) 

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Выводы**

Пропусков в колоках больше нет, они заменены на медиану.

### Изменение типов данных.

In [26]:
# Используем функцию astype для изменения типа данных на целочисленный в колонке total_income
df['total_income'] = df['total_income'].astype(int) 

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


**Выводы**

Поменяли тип данных в колонке `total_income`. Теперь данные смотрятся лучше.

### Удаление дубликатов.

In [28]:
# Посмотрим уникальные значения в колонке education
display(df['education'].unique()) 

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Видны не явнвные дубликаты, которые обычно появляются из-за человеческого фактора. Кот-то пишет в нижнем регистре, кто-то в верхнем.

In [29]:
# Сделаем все записи в колонке education в нижнем регистре. С помощью метода str.lower()
df['education'] = df['education'].str.lower() 

In [30]:
# Посмотрим еще раз уникальные значения в колонке education
display(df['education'].unique()) 

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [31]:
# Посмотрим уникальные значения в family_status
display(df['family_status'].unique()) 

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [32]:
# Сделаем все записи в колонке family_status в нижнем регистре. С помощью метода str.lower()
df['family_status'] = df['family_status'].str.lower() 

In [33]:
# Посмотрим еще раз уникальные значения в колонке family_status
display(df['family_status'].unique()) 

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

In [34]:
# Посмотрим количество дубликатов
df.duplicated().sum() 

71

In [35]:
# Посмотрим на них в таблице
display(df[df.duplicated()==True]) 

children  days_employed  dob_years education  education_id  \
2849          0    2194.220567         41   среднее             1   
3290          0    2194.220567         58   среднее             1   
4182          1    2194.220567         34    высшее             0   
4851          0    2194.220567         60   среднее             1   
5557          0    2194.220567         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0    2194.220567         64   среднее             1   
21032         0    2194.220567         60   среднее             1   
21132         0    2194.220567         47   среднее             1   
21281         1    2194.220567         30    высшее             0   
21415         0    2194.220567         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         145017            покупка жилья для семьи  
3290         145017                    сыграть свадьбу  
4182         145017                            свадьба  
4851         145017                            свадьба  
5557         145017                    сыграть свадьбу  
...             ...                                ...  
20702        145017         дополнительное образование  
21032        145017              заняться образованием  
21132        145017                       ремонт жилью  
21281        145017  покупка коммерческой недвижимости  
21415        145017                  операции с жильем  

[71 rows x 12 columns]

In [36]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True) 

In [37]:
# Посмотрим еще раз количество дубликатов
display(df.duplicated().sum()) 

0

**Выводы**

Перезаписали все значения в колонках `education` и `family_status` на нижний регистр. Так же удалили явные дубликаты из датафрейма

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [38]:
# Создадим новый DataFrame (df_education) с уникальными значениями колонок education и education_id,
# при помощи методов .drop_duplicates() и .reset_index 
df_education = df[['education', 'education_id']].drop_duplicates().reset_index(drop=True) 

In [39]:
# Создадим новый DataFrame (df_family_status) с уникальными значениями колонок family_status и family_status_id,
# при помощи методов .drop_duplicates() и .reset_index 
df_family_status = df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True) 

In [40]:
# Посмотрим что получилось в df_education
display(df_education) 

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [41]:
# Посмотрим что получилось в df_education
display(df_family_status) 

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

In [42]:
# Удалим колонки education и family_status из исходного дата фрейма
df = df.drop(columns=['education', 'family_status']) 

**Выводы**

Создали два новых дата фрейма: `df_education` и `df_family_status`. И удалили колонки `education` `family_status` из исходного дата фрейма.

### Категоризация дохода.

На основании диапазонов `total_income`, создадим столбец `total_income_category` с категориями.

In [43]:
# Функция возвращает группу по значению total_income — ежемесячный доход, используя правила:
def total_income_category(total_income): 
    try:
        if total_income <= 30000:            # - 'E', если total_income <= 30000;                          
            return 'E'                       # - 'D', если total_income <= 50000;
        if total_income <= 50000:            # - 'C' если total_income <= 200000;
            return 'D'                       # - 'B' если total_income <= 1000000;
        if total_income <= 200000:           # - 'A' если total_income от 1000001 и выше.
            return 'C'                       # обернем все проверки в конструкцию try except на случай ошибки в данных
        if total_income <= 1000000:
            return 'B'
        return 'A'
    except:
        display('Это число не подходит. Введите ')
        
        
        
        
        
        
    
   

In [44]:
# Создадим новый столбец (total_income_category), для этого нужен метод apply(): 
# он берёт значения столбца датафрейма и применяет к ним функцию из своего аргумента
df['total_income_category'] = df['total_income'].apply(total_income_category) 

In [45]:
# Посмотрим первые 10 строк таблицы
display(df.head(10)) 

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

**Выводы**

Написали функцию, которая возвращает группу по значению `total_income` и создали столбец с этой группой (`A`, `B`, `C`, `D`, `E`) в нашем датафрейме.

### Категоризация целей кредита.

На основании диапазонов `purpose`, создадим столбец `purpose_category` с категориями.

In [46]:
# Посмотрим все уникальные значения в столбце purpose
df['purpose'].unique() 

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [47]:
def purpose_category(purpose):                         # Функция возвращает группу по значению purpose — цель получения кредита, используя правила:
    try:
        if 'авто' in purpose:                          # - 'операции с автомобилем', если purpose содержит авто;
            return 'операции с автомобилем'            # - 'операции с недвижимостью', если purpose содержит 'недвиж' или 'жил';
        if 'жиль' in purpose or 'недвиж' in purpose:   # - 'проведение свадьбы', если purpose содержит свадьб;
            return 'операции с недвижимостью'          # - 'получение образования', если purpose содержит образ;
        if 'свадьб' in purpose:                        # обернем все проверки в конструкцию try except на случай ошибки в данных.
            return 'проведение свадьбы'
        if 'образ' in purpose:
            return 'получение образования'
    except:
        display('Что-то пошло не так. Это не подходящие данные.')
        
        
        
        
        
        
        
        
        
        
    

In [48]:
# Создадим новый столбец (purpose_category), для этого нужен метод apply(): 
# он берёт значения столбца датафрейма и применяет к ним функцию из своего аргумента
df['purpose_category'] = df['purpose'].apply(purpose_category) 

In [49]:
# Посмотрим уникальные значения столбца purpose_category
display(df['purpose_category'].unique()) 

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

**Выводы**

Написали функцию, которая возвращает группу по значению `purpose` и создали столбец с этой группой (`операции с автомобилем`, `операции с недвижимостью`, `проведение свадьбы`, `получение образования`) в нашем датафрейме.

### Ответы на вопросы.

#### Вопрос: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Сделаем сводную табилцу по количеству детей и задолженности по возврату кредитов.

In [50]:
# Используем метод pivot_table() и посмотрим сводную таблицу по числу детей и колонке debt, где 1 означает,
# что имеется задолженность по возврату кредитов, а 0 что не имеется.
display(df.pivot_table(index=['children'], columns=['debt'], values=['income_type'], aggfunc='count', fill_value=0)) 

income_type      
debt               0     1
children                  
0              13028  1063
1               4410   445
2               1926   202
3                303    27
4                 37     4
5                  9     0

In [51]:
# посмотрим среднее значение (ratio)
display(df.pivot_table(index=['children'], values=['debt'])) 

debt
children          
0         0.075438
1         0.091658
2         0.094925
3         0.081818
4         0.097561
5         0.000000

**Выводы**

Из этих данных можно сделать вывод, что клиенты без детей отвественнее всего подходят к выплате кредита в срок.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [52]:
# Объединим два дата фрейма df и df_family_status
data_family_status = df.merge(df_family_status, on='family_status_id', how='left') 

In [53]:
# Используем метод pivot_table() и посмотрим сводную таблицу по семейному положению и колонке debt, где 1 означает,
# что имеется задолженность по возврату кредитов, а 0 что не имеется.
display(data_family_status.pivot_table(index=['family_status'], columns=['debt'], values=['income_type'], aggfunc='count', fill_value=0)) 

income_type     
debt                            0    1
family_status                         
в разводе                    1110   85
вдовец / вдова                896   63
гражданский брак             3763  388
женат / замужем             11408  931
не женат / не замужем        2536  274

In [54]:
# посмотрим среднее значение (ratio)
display(data_family_status.pivot_table(index=['family_status'], values=['debt'])) 

debt
family_status                  
в разводе              0.071130
вдовец / вдова         0.065693
гражданский брак       0.093471
женат / замужем        0.075452
не женат / не замужем  0.097509

**Выводы**

Из таблицы данных видно, что вдовец/вдова самые ответственные из клиентов.

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [55]:
# Используем метод pivot_table() и посмотрим сводную таблицу по категории дохода и колонке debt, где 1 означает,
# что имеется задолженность по возврату кредитов, а 0 что не имеется.
display(df.pivot_table(index=['total_income_category'], columns=['debt'], values=['income_type'], aggfunc='count', fill_value=0)) 

income_type      
debt                            0     1
total_income_category                  
A                              23     2
B                            4685   356
C                           14656  1360
D                             329    21
E                              20     2

In [56]:
# посмотрим среднее значение (ratio)
display(df.pivot_table(index=['total_income_category'], values=['debt'])) 

debt
total_income_category          
A                      0.080000
B                      0.070621
C                      0.084915
D                      0.060000
E                      0.090909

**Выводы**

Лучший средний показатель у клиентов из категрии `D`. Клиенты с небольшим доходом знают, что в им нужна хорошая кредитная история.

#### Как разные цели кредита влияют на его возврат в срок?

In [57]:
# Используем метод pivot_table() и посмотрим сводную таблицу по целям кредита и колонке debt, где 1 означает,
# что имеется задолженность по возврату кредитов, а 0 что не имеется.
df.pivot_table(index=['purpose_category'], columns=['debt'], values=['income_type'], aggfunc='count', fill_value=0) 

income_type     
debt                               0    1
purpose_category                         
операции с автомобилем          3903  403
операции с недвижимостью       10029  782
получение образования           3643  370
проведение свадьбы              2138  186

In [58]:
# посмотрим среднее значение (ratio)
df.pivot_table(index=['purpose_category'], values=['debt']) 

debt
purpose_category                  
операции с автомобилем    0.093590
операции с недвижимостью  0.072334
получение образования     0.092200
проведение свадьбы        0.080034

**Выводы**

Луший показатель у клиентов с целью покупки недвижимости или жилья.

## Общий вывод:

Клиенты без детей отвественнее всего подходят к выплате кредита в срок. А если смотреть семейное положение то вдовец/вдова самые ответственные из клиентов, но их не так много от общего числа клиентов. У клиентов с целью покупки недвижимости или жилья тоже показатель очень высокий. Лучший средний показатель у клиентов из категрии `D`. Клиенты с небольшим доходом знают, что в им нужна хорошая кредитная история.